Add Github Actions jobs for testing the Rust API in CI

NlightNFotis opened the pull request:
This PR is in-progress. It is adding two more jobs, which are specialised to just build CBMC and then link+test the Rust API in two environments: a MacOS based one and a Linux based one.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add an Ubuntu 22.04 and MacOS 12 based Test Runner for testing the Rust API.

These two jobs exist solely for building CBMC and then linking and testing
the Rust API. They explicitly avoid running all of the tests for CBMC, as
we have dedicated jobs for doing just that.

NlightNFotis committed:
Change C smoke-test file path so the tests don't fail trying to open the file

thomasspriggs commented:
Why are we adding entirely new jobs for this rather than switching the flag on for some of the existing jobs?

NlightNFotis commented:
@thomasspriggs I had a few reasons in mind:

* I wanted to have complete control over the job scripts (to lean them out - control build configuration without interfering with other jobs, be able to install more software if I need to).
* To run them in parallel to the other jobs because these don't need to run the normal regression test suite, which allows for less time spent running the job and a cleaner log.
* To keep the knowledge for the Rust API in CI in a different (i.e. less bloated, cleaner) build configuration.
* The exact build configuration I wanted (CMake + Clang + Ubuntu 22.04) wasn't something that was pre-existing in the other CI array, so I would have to make a new job there in any case if I were to test within the environment I wanted (or hope I don't break another one massively).

Is there any reason for why you find the current configuration notable?

thomasspriggs commented on src/libcprover-rust/other/example.c:
```diff
@@ -0,0 +1,7 @@
+// File just to allow smoke testing of Rust API while tests are run.
```
Shouldn't test code be in the `regression` or `unit` directories? `other` seems like a particularly unhelpful categorisation.

NlightNFotis replied:
I agree that `other/` seems a bit misnamed (will probably change to something else as part of another PR), but I don't think this code should be in either `regression/` or `unit/` folders, on the basis that this is a small smoke test that is a dependency of the testing of the Rust API. 

All of the testing of the Rust API is happening inside the Rust project, and needs to be independent (as much as it can) and have all of its dependencies around so that a testing run is logically coherent and maintainable.

tautschnig replied:
But then we could have a `test` folder alongside `src`, couldn't we?

thomasspriggs reviewed (approved):
I am approving as I think it is best to get this merged and worry about any further details later, rather than leaving this work untested in CI.

Is there no Windows job due to it not working on Windows?

NlightNFotis commented:
I haven't tested it on Windows yet - this is probably going to be another ticket all together.

This is assuming that the Rust API working on windows is a goal. For now, I have focused on getting the functionality to work on MacOS (which is the machines that I'm using) and Linux (on the basis of the port being easy, and validated by the tests in this CI array).

NlightNFotis commented:
I am investigating some Windows CI networking issues which seem to be consistent to this job? (I don't know how - all other jobs are passing and this one has no change to Windows CI - or any of the pull request jobs for that matter).

I have rebased on top of the most recent `develop` branch and waiting for CI checks to complete, in case it is a transient issue with the Windows runner on that particular job.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+        run: z3 --version
+      - name: Log cargo/rust version
+        run: cargo --version
+      - name: Prepare ccache
+        uses: actions/cache@v3
+        with:
+          path: .ccache
+          key: ${{ runner.os }}-Release-${{ github.ref }}-${{ github.sha }}-PR
+          restore-keys: |
+            ${{ runner.os }}-Release-${{ github.ref }}
+            ${{ runner.os }}-Release
```
Please make either this one (or, perhaps better, the existing `check-macos-12-cmake-clang` test) use a different cache key name: the existing test uses `-Dsat_impl=glucose`, which will mean the compiler command lines are quite different. Therefore, they will be trashing each other's cache.

NlightNFotis replied:
I have changed it now to 

```
key: ${{ runner.os }}-Release-${{ github.ref }}-${{ github.sha }}-PR-Rust-API
```

Is this better?

tautschnig replied:
No, the prefix must not overlap with other jobs accepting the same prefix. Maybe the other job should say Release-glucose in the middle of the string.

NlightNFotis replied:
I have now changed the key for the job in [.github/workflows/pull-request-checks.yaml](https://github.com/diffblue/cbmc/pull/7498#diff-592c24d20f69c87f5c8c0450bf73355b6cf7ec646df4592ea10580f01e833704) to include the `Release-Glucose` tag so that it doesn't get mixed up with the cache keys for the Rust API build.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq clang-13 clang++-13 gdb maven jq flex bison libxml2-utils cpanminus ccache z3
+          make -C src minisat2-download
+          cpanm Thread::Pool::Simple
```
While I'm asking for changes: this line AFAIK has no useful effect (other than producing a warning).

NlightNFotis replied:
Yeah, I was planning on doing a second pass over this file in another PR as soon as I got stabilised enough to be merged, but I guess I can clean it up now as well.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+          DEBIAN_FRONTEND: noninteractive
+        run: |
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq clang-13 clang++-13 gdb maven jq flex bison libxml2-utils cpanminus ccache z3
```
I don't think you need `cpanminus`, `z3`, `libxml2-utils`, `jq`, `gdb` here.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+        run: |
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq clang-13 clang++-13 gdb maven jq flex bison libxml2-utils cpanminus ccache z3
+          make -C src minisat2-download
```
You're using `cmake`, this is useless.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+          sudo apt-get install --no-install-recommends -yq clang-13 clang++-13 gdb maven jq flex bison libxml2-utils cpanminus ccache z3
+          make -C src minisat2-download
+          cpanm Thread::Pool::Simple
+      - name: Confirm z3 solver is available and log the version installed
+        run: z3 --version
```
Won't be used, so don't needs to be installed.

NlightNFotis replied:
Was thinking about this in particular that it could be used to test against in the future, but I suppose I'm not going to need it yet, so better to keep it clean.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+      - name: Zero ccache stats and limit in size
+        run: ccache -z --max-size=500M
+      - name: Configure using CMake
+        run: cmake -S. -B${{env.default_build_dir}} -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DCMAKE_C_COMPILER=/usr/bin/clang-13 -DCMAKE_CXX_COMPILER=/usr/bin/clang++-13 -DWITH_RUST_API=ON
```
Could you please explain _why_ you need PIC? The commit message, unfortunately, lacks a body (which I would have hoped to explain this move).

NlightNFotis replied:
The simple answer is that CI failed without it (reproducibly on CI, claims that it needs everything recompiled with `-fPIE`.

Worth mentioning that Enrico couldn't reproduce locally on his machine, but given that this got the build to complete and do the actual Rust test run, I'm happy to leave this in for this particular configuration (doesn't seem to do any harm).

tautschnig replied:
No harm is fine, but the reasons should be documented in some form.

NlightNFotis replied:
This has now been documented in the form of a documentation comment in the same action script, explaining the need for the flag.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Fetch dependencies
+        run: brew install cmake ninja maven flex bison ccache z3
+      - name: Confirm z3 solver is available and log the version installed
+        run: z3 --version
```
As above: you don't actually need `z3`, do you?

NlightNFotis replied:
As above, we don't - I removed.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+      - name: Zero ccache stats and limit in size
+        run: ccache -z --max-size=500M
+      - name: Configure using CMake
+        run: cmake -S. -Bbuild -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DWITH_RUST_API=ON
```
Use `-B${{env.default_build_dir}}`, else the environment variable is pointless.

NlightNFotis replied:
Good catch - the environment variable is not pointless without it (it's primarily there as a reference to the build folder for the linking step of the Rust build) but yeah it should be referenced here as well.

tautschnig commented on .github/workflows/pull-request-check-rust-api.yaml:
```diff
@@ -0,0 +1,105 @@
+      - name: Configure using CMake
+        run: cmake -S. -Bbuild -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DWITH_RUST_API=ON
+      - name: Build with Ninja
+        run: cd build; ninja -j3
```
Need to use `${{env.default_build_dir}}`

NlightNFotis commented:
Hi @tautschnig I have addressed all of your review comments in the latest commit - [6618647](https://github.com/diffblue/cbmc/pull/7498/commits/661864741b78a30fe3d21e51a0adc72d60110943)

Is there anything else I should respond to - or is this good to go?
